versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/TracingDatabaseAdapter.java

```diff
@@ -0,0 +1,274 @@
+      return delegate.namedRefs(params).onClose(trace::close);
+    } catch (RuntimeException | ReferenceNotFoundException e) {
+      trace.close();
+      throw e;
+    }
+  }
+
+  @Override
+  public Hash create(NamedRef ref, Hash target)
+      throws ReferenceAlreadyExistsException, ReferenceNotFoundException {
+    try (Traced ignore =
+        trace("create").tag(TAG_REF, ref.getName()).tag(TAG_HASH, target.asString())) {
+      return delegate.create(ref, target);
+    }
+  }
+
+  @Override
+  public void delete(NamedRef reference, Optional<Hash> expectedHead)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    try (Traced ignore = trace("delete").tag(TAG_REF, reference.getName())) {
+      delegate.delete(reference, expectedHead);
+    }
+  }
+
+  @Override
+  public void assign(NamedRef assignee, Optional<Hash> expectedHead, Hash assignTo)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    try (Traced ignore =
+        trace("assign").tag(TAG_HASH, assignTo.asString()).tag(TAG_REF, assignee.getName())) {
+      delegate.assign(assignee, expectedHead, assignTo);
+    }
+  }
+
+  @Override
+  public Stream<Difference> diff(Hash from, Hash to, KeyFilterPredicate keyFilter)
+      throws ReferenceNotFoundException {
+    Traced trace = trace("diff").tag(TAG_FROM, from.asString()).tag(TAG_TO, to.asString());
+    try {
+      return delegate.diff(from, to, keyFilter).onClose(trace::close);
+    } catch (RuntimeException | ReferenceNotFoundException e) {
+      trace.close();
+      throw e;
+    }
+  }
+
+  @Override
+  public RepoDescription fetchRepositoryDescription() {
+    try (Traced ignore = trace("fetchRepositoryDescription")) {
+      return delegate.fetchRepositoryDescription();
+    }
+  }
+
+  @Override
+  public void updateRepositoryDescription(Function<RepoDescription, RepoDescription> updater)
+      throws ReferenceConflictException {
+    try (Traced ignore = trace("updateRepositoryDescription")) {
+      delegate.updateRepositoryDescription(updater);
+    }
+  }
+
+  @Override
+  public Stream<ContentIdWithType> globalKeys(ToIntFunction<ByteString> contentTypeExtractor) {
+    Traced trace = trace("globalKeys");
+    try {
+      return delegate.globalKeys(contentTypeExtractor).onClose(trace::close);
+    } catch (RuntimeException e) {
+      trace.close();
+      throw e;
+    }
+  }
+
+  @Override
+  public Stream<ContentIdAndBytes> globalContent(
+      Set<ContentId> keys, ToIntFunction<ByteString> contentTypeExtractor) {
+    Traced trace = trace("globalContent").tag(TAG_COUNT, keys.size());
+    try {
+      return delegate.globalContent(keys, contentTypeExtractor).onClose(trace::close);
+    } catch (RuntimeException e) {
+      trace.close();
+      throw e;
+    }
+  }
+
+  @Override
+  public Optional<ContentIdAndBytes> globalContent(
+      ContentId contentId, ToIntFunction<ByteString> contentTypeExtractor) {
+    try (Traced ignore = trace("globalContent")) {
```

Comment:
should we add the `contentId` as a tag here?

Reply:
done